org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java

```diff
@@ -347,80 +383,125 @@
         } else if (opt.solver.equals(A4Options.SatSolver.MiniSatProverJNI)) {
             sym = 20;
-            solver.options().setSolver(SATFactory.MiniSatProver);
-            solver.options().setLogTranslation(2);
-            solver.options().setCoreGranularity(opt.coreGranularity);
+            solver_opts.setSolver(SATFactory.MiniSatProver);
+            solver_opts.setLogTranslation(2);
+            solver_opts.setCoreGranularity(opt.coreGranularity);
         } else {
-            solver.options().setSolver(SATFactory.DefaultSAT4J); // Even for
-                                                                // "KK" and
-                                                                // "CNF", we
-                                                                // choose
-                                                                // SAT4J
-                                                                // here;
-                                                                // later,
-                                                                // just
-                                                                // before
-                                                                // solving,
-                                                                // we'll
-                                                                // change it
-                                                                // to a
-                                                                // Write2CNF
-                                                                // solver
+            // Even for "KK" and "CNF", we choose SAT4J here; later, just before solving, we'll change it to a Write2CNF solver
+            solver_opts.setSolver(SATFactory.DefaultSAT4J);
+        }
+        solver_opts.setSymmetryBreaking(sym);
+        solver_opts.setSkolemDepth(opt.skolemDepth);
+        solver_opts.setBitwidth(bitwidth > 0 ? bitwidth : (int) Math.ceil(Math.log(atoms.size())) + 1);
+        solver_opts.setIntEncoding(Options.IntEncoding.TWOSCOMPLEMENT);
+        // [HASLab] create unique readable name
+        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd-HH-mm");
+        String file = "untitled";
+        if (!getOriginalFilename().isEmpty()) {
+            String[] pths = getOriginalFilename().split("/");
+            file = pths[pths.length - 1].substring(0, pths[pths.length - 1].length() - 4).replace(' ', '_');
         }
-        solver.options().setSymmetryBreaking(sym);
-        solver.options().setSkolemDepth(opt.skolemDepth);
-        solver.options().setBitwidth(bitwidth > 0 ? bitwidth : (int) Math.ceil(Math.log(atoms.size())) + 1);
-        solver.options().setIntEncoding(Options.IntEncoding.TWOSCOMPLEMENT);
+        String check = getOriginalCommand().replace(' ', '_').replace('$', '-');
+        solver_opts.setUniqueName(file + "-" + check + "-" + dateFormat.format(new Date()) + "-" + this.hashCode());
```

Comment:
solving at kodkod only considers a formula and bounds, this allows some kind of traceability for debugging purposes